### ISSUE TITLE
[Feature request] - Add a hystereris on `over_switch` and `over_valve` in complement of the TPI algorithm

### ISSUE THREAD

**Issue by jmcollin78** [jmcollin78/versatile_thermostat#716]
See:
```
I imagined a command in over_switch or over_valve.
The regulation could be done as follows:

regulation in TPI until obtaining the value of T°C setpoint + hysteresis
we cut the regulation
as soon as the value read reaches T°C setpoint - hysteresis, we restart the cycle.
```

Vote for this FR if you are interested.

**Comment by mcornillon** [comment]
And I'd like to have it also for the "over_climate with direct valve control".
That would be very interesting, notably for installations with a central boiler, so that the boiler is started only when a significant amount of heat needs to be generated.

And it would be nice to have hysteris configurable, and differentiated between the "+ hysteris" from the "- hysteresis" (could be +0.2°, but -0.5° for example)

**Comment by jmcollin78** [comment]
Hello implementation will starts due to the large number of interested users.

VTherm will start TPI regulation only if current_temp < target + hysteresis_low
VTherm will stop TPI regulation only  if current_temp > target + hysteresis_high

For target_temp = 20°, setting hysteresis_low to 0.2 and hysteresis_high to 0.5, means that TPI will start switching on/off the heater is current_temp is < 20.2 (20+0.2) and stop the TPI algo when current_temp is > 20.5 (20+0.5).

Setting the values to 0 will behave like today.

Note that this is somehow in contradiction with the min activation delay.
<img width="209" height="73" alt="Image" src="https://github.com/user-attachments/assets/f89bde2b-9b1c-4fcb-8e55-9afa86699ab2" />
 The goal of the min activation delay is to avoid too short delay switching On or off. This is also a kind of hystereris. So the both conditions (min activation delay and hysteresis) should be satisfied to turn on the TPI algorithm.

Is that ok for you ?

EDIT: @mcornillon, it will works also for over climate with valve regulation because it uses also the TPI algorithm.

**Comment by jmcollin78** [comment]
Please give a try to this pre-release https://github.com/jmcollin78/versatile_thermostat/releases/tag/7.4.0.beta1
Don't hesitate to give feedback if you need it.

Documentation has also been updated in the 716-xx branch.

**Comment by mcornillon** [comment]
Thank you @jmcollin78 for this update.
Unfortunately I won't be able to test it and provide feedback, I've left France 1 month ago, and now live 10 000 kms away in a new country, where no heating is required (but might try VTherm one day with our bunch of air conditioners) !
Good luck for this coming winter and all new PR
And once again, congrats for your great project

**Comment by jmcollin78** [comment]
Result of my own test:
1. initial condition (no heating needed):
<img width="447" height="590" alt="Image" src="https://github.com/user-attachments/assets/1b3b3dfb-6809-4d65-9d47-55b2d6fec0ce" />

<img width="427" height="273" alt="Image" src="https://github.com/user-attachments/assets/d499b7b5-f5d7-41b4-8ea3-27498c503725" />

## Test the low threshold: when current_temp decrease, the VTherm heats when current_temp - target < low threshold
2. Temp slow down from 22° (target = 20°) to 19° with 0.1 step,
Each time I see these log:
`Proportional algorithm: on_percent is forced to 0 cause current_temp (21.7) is outside the thresholds (slope=-8.1, target_temp=20.0, tpi_threshold_low=1.0, tpi_threshold_high=2.0). Heating/cooling will be disabled`

3. at current_temp = 21° the VTherm starts to heat (20°+ low threshold = 21°)
See ont_time which starts at 30s and the cycle which starts at this time to:
<img width="436" height="887" alt="Image" src="https://github.com/user-attachments/assets/84459fb6-49f2-4a8e-a7b6-fca30b6c358a" />

4. at current_temp = 19° here is the situation:

<img width="434" height="888" alt="Image" src="https://github.com/user-attachments/assets/bde9df62-ec06-4ce9-8100-32d87a5bdcc6" />

## Test the hogh threshold: when current_temp increase, the VTherm stop heating when current_temp - target > high threshold
5.  Temp increase slowly from 19° to 22,5°. At 22° it should stop heating
At 20.5°, the slope is now posittive, the on percent is 45%, the thermostat continue to heat at its cycle:

<img width="434" height="894" alt="Image" src="https://github.com/user-attachments/assets/79cbbcc5-74f0-46ab-b35d-3f08e9bab7ca" />

6. Above 21,1°, the VTher continue to heat. We have now the on percent = 26%

<img width="434" height="894" alt="Image" src="https://github.com/user-attachments/assets/2e05a634-9ec5-4188-898b-f8af01a8ae32" />

7. At 22.1° the VTherm stop heating. (Note that I have to change the ext temp because at 22° on percent is 0 without threshold)
8. I see the log: `Thermostat switch 1 - Proportional algorithm: on_percent is forced to 0 cause current_temp (22.1) is outside the thresholds (slope=11.0, target_temp=20.0, tpi_threshold_low=1.0, tpi_threshold_high=2.0). Heating/cooling will be disabled.`

The VTherm stop heating, the on percent is 0 (due to threshold thanks to the log):
<img width="426" height="887" alt="Image" src="https://github.com/user-attachments/assets/803f5fd6-5da7-4112-80a4-ee266061f138" />

# Conclusion: this works as expected.

**Comment by Liborsaf** [comment]
Hello @jmcollin78, thanks a lot for implementing this feature!

I'm currently trying to set it up, but it doesn't seem to work for me.
My configuration is:
- Setpoint: 22 °C
- Upper threshold: 2 °C
- Lower threshold: 1 °C

From my understanding, the behavior should be:
- Heating starts when the temperature drops to 20 °C (setpoint - lower threshold).
- Heating stops when the temperature reaches 23 °C (setpoint + upper threshold).

But it doesn't appear to be working as expected for me.
Could you please help me figure out if I'm missing something in the setup?

**Comment by Liborsaf** [comment]
I got it working now, I just forgot to restart Home Assistant due to #1025.

However, right now the lower threshold seems to work as setpoint + lower threshold to start heating, instead of setpoint - lower threshold.

Would it be possible to add an option to allow negative values for lower threshold?

**Comment by jmcollin78** [comment]
Thank you for testing it !

Negative values for lower threshold doesn't make sense for me. It is obviously possible to allow it but you will then have a current temperature  < the target temperature. Is that really what you want ?
All in VTherm is done to have the current temp near the target. What is the use case ? What do you want to be cold ?

**Comment by Liborsaf** [comment]
I want to use this to control a pellet stove as a heater. To avoid multiple on/off toggles, I want the heating to start only once the temperature drops a bit below my setpoint. I do not really mind being 2 °C under the setpoint before it starts heating again.

This most likely would not impact the existing logic much, it would just add more flexibility for users if negative values were allowed. I believe that is also what @mcornillon had in mind.

**Comment by jmcollin78** [comment]
Got it. I publish a new beta soon with that point
.

**Comment by jmcollin78** [comment]
Done in https://github.com/jmcollin78/versatile_thermostat/releases/tag/7.4.0.beta2

**Comment by Liborsaf** [comment]
Thanks for your time with this.

I just tried it, and it seems it still doesn't work. With a negative value it stops working completely. After looking into the source code I think this could be the issue:
https://github.com/jmcollin78/versatile_thermostat/blob/a8d2a90e530efd834257ad14e1f48bcf38cd69df/custom_components/versatile_thermostat/prop_algorithm.py#L94

**Comment by jmcollin78** [comment]
I will add some test with negatives values because I suppose that 0 means no value (default value)

**Comment by jmcollin78** [comment]
This times it works at home (on my test environment): https://github.com/jmcollin78/versatile_thermostat/releases/tag/7.4.0.beta3

**Comment by Liborsaf** [comment]
I tested it and it seems to work perfectly now. Thank you so much! 😄

**Comment by jmcollin78** [comment]
Thank you for your test and report. I will release it next week.

**Comment by adi90x** [comment]
Working as expected, but in real world isn't it more effective to play a bit on the ``coeff_ext`` param to stop TPI algorithm to heat as this is supposed to be used to control the long term trend, no ?

**Comment by jmcollin78** [comment]
> Working as expected, but in real world isn't it more effective to play a bit on the `coeff_ext` param to stop TPI algorithm to heat as this is supposed to be used to control the long term trend, no ?

Thank you for the suggestion but I think it is false good idea. Touching the `coeff_ext` will not ensure you to stop heating and will only make the target temp not reach. This is not what we want. We want to oscillate around the target temp with hysteresis.

**Comment by jmcollin78** [comment]
https://github.com/jmcollin78/versatile_thermostat/releases/tag/7.4.0
